### PR TITLE
Play until bindings and CLI

### PR DESCRIPTION
**Pull request by agalbachicar**
Goes on top of #4 

Should review f44a729f03a07f2af4d80fd1610aa12ed2b10696

Can use by doing:

```sh
ros2 bag_bp play -u 123.456 bag_file
ros2 bag_bp --until 123.456 bag_file
```

When the parameter is omitted, default behavior (unconstrained by time) is performed.

Note that `123.456` is measured in seconds, an internal conversion to nanoseconds is done at the Python layer.

### CODE REVIEW

**Comment by agalbachicar** [comment]
@gbiggs  for review.

File: ros2bag/ros2bag_backport/verb/play.py [hunk 1]
```diff
@@ -120,6 +129,18 @@ def main(self, *, args):  # noqa: D102
...
         play_options.delay = args.delay
+        # Gets the duration in nanoseconds when a value is provided for player
+        # consumption.
+        duration = int(args.duration * 1e9) if args.duration else args.duration
+        # Gets the timestamp in nanoseconds when a value is provided for player
+        # consumption.
+        timestamp = int(args.until * 1e9) if args.until else args.until
+        if duration and timestamp:
+            raise AssertionError('Must provide either duration, timestamp or '
+                                 'nothing but not both at the same time.')
```
Comment (gbiggs) [inline ros2bag/ros2bag_backport/verb/play.py]:
```suggestion
            raise AssertionError('Duration and timestamp cannot be used together.')
```

File: rosbag2_transport/src/rosbag2_transport/player.cpp [hunk 2]
```diff
@@ -228,6 +242,25 @@ bool Player::is_storage_completely_loaded() const
...
+}
+
+void Player::do_play(
+  const std::optional<rcutils_duration_value_t> & duration,
+  const std::optional<rcutils_time_point_value_t> & timestamp)
+{
+  // This is a faulty condition. This method must be called exclusively with
+  // one or none of the attributes set, but not both.
+  if (duration.has_value() && timestamp.has_value()) {
+    RCLCPP_ERROR(this->get_logger(), "Failed to play. Both duration and until timestamp are set.");
```
Comment (gbiggs) [inline rosbag2_transport/src/rosbag2_transport/player.cpp]:
```suggestion
    RCLCPP_ERROR(this->get_logger(), "Failed to play. Both duration and 'until' timestamp are set.");
```

File: rosbag2_transport/test/rosbag2_transport/test_play_until.cpp [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+    auto primitive_message1 = get_messages_basic_types()[0];
+    primitive_message1->int32_value = kIntValue;
+
+    auto complex_message1 = get_messages_arrays()[0];
+    complex_message1->float32_values = {{kFloat1Value, kFloat2Value, kFloat3Value}};
+    complex_message1->bool_values = {{kBool1Value, kBool2Value, kBool3Value}};
+
+    // @{ Ordering matters. The mock reader implementation moves messages
+    //    around without any knowledge about message chronology. It just picks
+    //    the next one Make sure to keep the list in order or sort it!
```
Comment (gbiggs) [inline rosbag2_transport/test/rosbag2_transport/test_play_until.cpp]:
```suggestion
    //    the next one. Make sure to keep the list in order or sort it!
```

File: rosbag2_transport/test/rosbag2_transport/test_play_until.cpp [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+  auto replayed_test_arrays = sub_->get_received_messages<test_msgs::msg::Arrays>(
+    kTopic2);
+  EXPECT_THAT(replayed_test_arrays, SizeIs(Eq(2u)));
+  EVAL_REPLAYED_BOOL_ARRAY_PRIMITIVES(replayed_test_arrays);
+  EVAL_REPLAYED_FLOAT_ARRAY_PRIMITIVES(replayed_test_arrays);
+}
+
+TEST_F(
+  RosBag2PlayUntilWithStartTimeOffsetTestFixture,
+  because_of_offset_none_message_is_received_for_all_topics)
```
Comment (gbiggs) [inline rosbag2_transport/test/rosbag2_transport/test_play_until.cpp]:
```suggestion
  because_of_offset_no_message_is_received_for_all_topics)
```

**Review by gbiggs — APPROVED** [review]
LGTM

**Comment by agalbachicar** [comment]
Feel free to go in!

**Comment by aprotyas** [comment]
Will this PR get upstreamed?

**Comment by gbiggs** [comment]
Yes, we plan on upstreaming it this month.

**Pull request merged**